src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,344 @@
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for Thread Link Metrics.
+ */
+
+#include "link_metrics.hpp"
+
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+
+#include "link_metrics_tlvs.hpp"
+
+namespace ot {
+
+LinkMetrics::LinkMetrics(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mLinkMetricsReportCallback(nullptr)
+    , mLinkMetricsReportCallbackContext(nullptr)
+{
+}
+
+otError LinkMetrics::LinkMetricsQuery(const Ip6::Address & aDestination,
+                                      uint8_t              aSeriesId,
+                                      const otLinkMetrics &aLinkMetricsFlags)
+{
+    otError                error;
+    LinkMetricsTypeIdFlags typeIdFlags[kMaxTypeIdFlags];
+    uint8_t                typeIdFlagsCount = GetTypeIdFlagsFromOtLinkMetricsFlags(typeIdFlags, aLinkMetricsFlags);
+
+    error = SendLinkMetricsQuery(aDestination, aSeriesId, typeIdFlags, typeIdFlagsCount);
+
+    return error;
+}
+
+otError LinkMetrics::AppendLinkMetricsReport(Message &                       aMessage,
+                                             const Mle::LinkMetricsQueryTlv *aLinkMetricsQuery,
+                                             const Message &                 aRequestMessage)
+{
+    otError                        error = OT_ERROR_NONE;
+    Tlv                            tlv;
+    const LinkMetricsQueryId *     queryId      = nullptr;
+    const LinkMetricsQueryOptions *queryOptions = nullptr;
+    uint8_t                        length       = 0;
+    uint16_t                       startOffset  = aMessage.GetLength();
+    const Tlv *                    subTlvCur    = nullptr;
+    const Tlv *                    subTlvEnd    = nullptr;
+
+    VerifyOrExit(aLinkMetricsQuery != nullptr && aLinkMetricsQuery->IsValid(), error = OT_ERROR_PARSE);
+    subTlvCur = aLinkMetricsQuery->GetSubTlvs();
+    subTlvEnd = aLinkMetricsQuery->GetNext();
+
+    // Parse QueryId Sub-TLV
+    if (subTlvCur->GetType() == kLinkMetricsQueryId)
+    {
+        queryId   = static_cast<const LinkMetricsQueryId *>(subTlvCur);
+        subTlvCur = subTlvCur->GetNext();
+        VerifyOrExit(queryId->IsValid(), error = OT_ERROR_PARSE);
+        VerifyOrExit(queryId->GetSeriesId() < 255, error = OT_ERROR_INVALID_ARGS);
```

Comment:
Name this `255` value in `link_metrics.hpp`?

Reply:
Actually this is valid in forward tracking series. So I removed this line.